[wip] add object index

jerryjliu opened the pull request:
# Description

Add ObjectIndex abstraction - allows serialization/deserialization from Node objects.

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [ ] I stared at the code and made sure it makes sense

logan-markewich commented on llama_index/indices/objects/base.py:
```diff
@@ -0,0 +1,58 @@
+    ) -> "ObjectIndex":
+        object_node_mapping = object_mapping_cls(objects)
+        nodes = object_node_mapping.to_nodes()
+        return index_cls(nodes, **index_kwargs)
```
This returns `index_cls`, rather than `cls(index, object_node_mapping)`, was that intended?

jerryjliu commented:
hm. not sure i like the form factor of the current ObjectNodeMapping class. going to refactor

jerryjliu commented on llama_index/objects/base_node_mapping.py:
```diff
@@ -0,0 +1,68 @@
+OT = TypeVar("OT")
+
+
+class BaseObjectNodeMapping(Generic[OT]):
```
cc @Disiok @logan-markewich i tried to make this interface a bit more general. I wasn't a fan of the base class taking in a list of objects by default. Since there could be cases (e.g. with the SQL table schema), where you can serialize/deserialize to/from Nodes without needing the list beforehand.

Storing a list of Tools helps because we don't have an easy way to serialize/deserialize to/from a Node/Tool.

jerryjliu reviewed (commented):
I still need to add notebook examples/tests

logan-markewich commented on llama_index/objects/base.py:
```diff
@@ -0,0 +1,61 @@
+    ) -> "ObjectIndex":
+        object_node_mapping = object_mapping_cls.from_objects(objects)
+        nodes = object_node_mapping.to_nodes()
+        return index_cls(nodes, **index_kwargs)
```
shouldn't this return `cls(index_cls(...), object_node_mapping)` ?

logan-markewich commented on llama_index/objects/base_node_mapping.py:
```diff
@@ -0,0 +1,68 @@
+
+    @classmethod
+    @abstractmethod
+    def from_objects(
```
Do you think there should be an `as_object_index()` method too? Or maybe that's getting too circular haha

hongyishi replied:
+1 since we already have the objects and the self NodeMapping. The user can pass in their own Index type or we can default to Vector Index.

jerryjliu replied:
Hm since the entry point is for the user to define an ObjectIndex as opposed to initializing a BaseNodeObjectMapping it would look like this:

```python
obj_index = ObjectIndex.from_objects(
    [list_tool, vector_tool], 
    VectorStoreIndex,
    SimpleToolNodeMapping
)
```

So IMO not a huge need to implement `as_object_index` atm

logan-markewich commented on llama_index/objects/base_node_mapping.py:
```diff
@@ -0,0 +1,68 @@
+
+
+class SimpleObjectNodeMapping(BaseObjectNodeMapping):
+    """General node mapping that works for any obj."""
```
"... any obj with a meaningful string representation" ? Just trying to imagine the use-case for this one (SQL tables seem like a good option!)

logan-markewich reviewed (approved):
Cool, I like it! Solves a common problem 💪🏻

hongyishi commented on llama_index/agent/retriever_openai_agent.py:
```diff
@@ -0,0 +1,90 @@
+from typing import Sequence, Dict
+
+
+# class OpenAIFnNodeMapping(BaseObjectNodeMapping[OpenAIFunctionSchema]):
```
Remove unused code

hongyishi commented on llama_index/objects/base.py:
```diff
@@ -0,0 +1,61 @@
+
+    def as_retriever(self, **kwargs) -> ObjectRetriever:
+        return ObjectRetriever(
+            retriever=self.as_retriever(**kwargs),
```
Not sure if self has a retriever. It should the index retriever?

jerryjliu replied:
fixed

hongyishi commented on llama_index/objects/base.py:
```diff
@@ -0,0 +1,61 @@
+        nodes = object_node_mapping.to_nodes()
+        return index_cls(nodes, **index_kwargs)
+
+    def insert_object(self, obj: Any):
```
Should we do some type checking that the inserted objects can be mapped?

jerryjliu replied:
that's a good point....looking into something quick right now

hongyishi commented on llama_index/objects/tool_node_mapping.py:
```diff
@@ -0,0 +1,68 @@
+
+    def from_node(self, node: Node) -> BaseTool:
+        """From node."""
+        return self._tools[node.text]
```
I don't think the _tools dict in the constructor uses the right keys in this case.

hongyishi commented on llama_index/objects/tool_node_mapping.py:
```diff
@@ -0,0 +1,68 @@
+
+    def from_node(self, node: Node) -> QueryEngineTool:
+        """From node."""
+        return self._tools[node.text]
```
Same as above

jerryjliu commented:
cc @logan-markewich @hongyishi @Disiok , i've done the following:
- added unit tests
- changed notebook for router and openai agent
- added validation checks to objects
